src/turbomind/kernels/sampling_topp_kernels.cu

```diff
@@ -27,795 +27,134 @@
-        case 2:
-            KERNEL_RUN(2) break;
-        case 3:
-            KERNEL_RUN(3) break;
-        case 4:
-            KERNEL_RUN(4) break;
-        case 5:
-            KERNEL_RUN(5) break;
-        case 6:
-            KERNEL_RUN(6) break;
-        case 7:
-            KERNEL_RUN(7) break;
-        default:
-            exit(1);
-    }
+    topPSortInitialize<<<32, 512, 0, stream>>>(
+        vocab_size_padded, vocab_size, batch_size, top_ks, topp_id_val_buf, begin_offet_buf, end_offset_buf);
 }
 
-///////////////////////////////////////////////////////////////////////////////////////////////////
-
-template<typename Kernel_params>
-void topPPerSegment_dispatch(const TopKPerSegmentContext& context,
-                             TopKPerSegmentParams&        params,
-                             void*                        temp_storage,
-                             size_t&                      temp_storage_bytes,
-                             cudaStream_t                 stream)
+template<typename T>
+static __global__ void softmax(T* logits, const int vocab_size_padded, const int* kept)
 {
+    int bid = blockIdx.x;
+    logits += bid * vocab_size_padded;
+    int n = kept[bid];
 
-    using Key_Data_Type   = typename Kernel_params::Key_Data_Type;
-    using Value_Data_Type = typename Kernel_params::Value_Data_Type;
+    float            max_val   = -1 * FLT_MAX;
+    const bool       IS_FP16   = std::is_same<T, half>::value;
+    const T          MAX_T_VAL = (IS_FP16) ? HALF_FLT_MAX : FLT_MAX;
+    __shared__ float s_max_val;
+    __shared__ float s_sum_val;
 
-    if (temp_storage == nullptr) {
-        if (params.num_segments > 1) {
-            cub::DeviceSegmentedRadixSort::SortPairsDescending(temp_storage,
-                                                               temp_storage_bytes,
-                                                               reinterpret_cast<Key_Data_Type*>(params.gmem_src_keys),
-                                                               reinterpret_cast<Key_Data_Type*>(params.gmem_dst_keys),
-                                                               reinterpret_cast<Value_Data_Type*>(params.gmem_src_vals),
-                                                               reinterpret_cast<Value_Data_Type*>(params.gmem_dst_vals),
-                                                               params.num_items,
-                                                               params.num_segments,
-                                                               params.gmem_begin_offsets,
-                                                               params.gmem_end_offsets,
-                                                               0,
-                                                               sizeof(Key_Data_Type) * 8,
-                                                               stream);
+    for (int tid = threadIdx.x; tid < vocab_size_padded; tid += blockDim.x) {
```

Comment:
Shall we skip this softmax for top-k request?

Reply:
removed